AX: Exiting early from AXObjectCache::childrenChanged(RenderObject*) when we can't `get` an object for anonymous renderers can result in content being missing from the accessibility tree

twilco opened the pull request:
#### 02c3f17b1e4a5a83009a8f897716a47c1413dbd4
<pre>
AX: Exiting early from AXObjectCache::childrenChanged(RenderObject*) when we can&apos;t `get` an object for anonymous renderers can result in content being missing from the accessibility tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=290069">https://bugs.webkit.org/show_bug.cgi?id=290069</a>
<a href="https://rdar.apple.com/147283521">rdar://147283521</a>

Reviewed by Chris Fleizach.

Exiting early from AXObjectCache::childrenChanged(RenderObject*) when we can&apos;t `AXObjectCache::get` an object for
anonymous renderers can result in content being missing from the accessibility tree. Sometimes the only children-changed
we get from the render tree for some subtrees is for an anonymous renderer, so if we just drop it, we can have a stale
accessibility tree. This problem is specific to anonymous renderers because we walk the DOM when building the accessibility
tree, so it&apos;s unlikely that we will have created an accessibility object for this renderer (since it only exists in the render tree).

This commit resolves the problem by changing AXObjectCache::childrenChanged(RenderObject*) to check whether a renderer
is anonymous, and if we can&apos;t AXObjectCache::get an accessibility object for it, ascends to the nearest thing we can get
an accessibility object for and submits the children-changed event for that instead.

This bug was found on a real tax preparation webpage, but unfortunately I couldn&apos;t figure out how to distill it into a layout test.
The key seems to be anonymous continuation renderers destroyed and recreated as part of a call to Node::insertBefore().
Newly added test dynamic-inline-continuation.html gets close to reproducing the issue, following many of the same
codepaths as the real bug, but unfortunately will still pass even without this commit.

For posterity, here&apos;s a stacktrace that gives us one of these children-changed events that used to be dropped but no
longer is:

2   AXObjectCache::childrenChanged(WebCore::RenderObject*, WebCore::RenderObject*)
3   RenderTreeBuilder::Block::detach(WebCore::RenderBlock&amp;, WebCore::RenderObject&amp;, WebCore::RenderTreeBuilder::WillBeDestroyed, WebCore::RenderTreeBuilder::CanCollapseAnonymousBlock)
4   RenderTreeBuilder::destroy(WebCore::RenderObject&amp;, WebCore::RenderTreeBuilder::CanCollapseAnonymousBlock)
5   RenderTreeBuilder::destroyAndCleanUpAnonymousWrappers(WebCore::RenderObject&amp;, WebCore::RenderElement const*)
6   RenderTreeUpdater::tearDownRenderers(WebCore::Element&amp;, WebCore::RenderTreeUpdater::TeardownType, WebCore::RenderTreeBuilder&amp;)
7   RenderTreeUpdater::tearDownRenderers(WebCore::Element&amp;, WebCore::RenderTreeUpdater::TeardownType)
8   ContainerNode::removeBetween(WebCore::Node*, WebCore::Node*, WebCore::Node&amp;)
9   ContainerNode::removeChild(WebCore::Node&amp;)
10  ContainerNode::removeSelfOrChildNodesForInsertion(WebCore::Node&amp;, WTF::Vector&lt;WTF::Ref&lt;WebCore::Node, WTF::RawPtrTraits&lt;WebCore::Node&gt;, WTF::DefaultRefDerefTraits&lt;WebCore::Node&gt;&gt;, 11ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc&gt;&amp;)
11  ContainerNode::appendChildWithoutPreInsertionValidityCheck(WebCore::Node&amp;)
12  ContainerNode::insertBefore(WebCore::Node&amp;, WTF::RefPtr&lt;WebCore::Node, WTF::RawPtrTraits&lt;WebCore::Node&gt;, WTF::DefaultRefDerefTraits&lt;WebCore::Node&gt;&gt;&amp;&amp;)

This commit also inlines AXObjectCache::get(RenderObject&amp;) and AXObjectCache::(Widget&amp;) for better performance.

* LayoutTests/accessibility/mac/dynamic-inline-continuation-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-inline-continuation.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::childrenChanged):
(WebCore::AXObjectCache::getIncludingAncestors const):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: <a href="https://commits.webkit.org/292433@main">https://commits.webkit.org/292433@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcc2693decc6c89fcaa43ad05fed14c9cdcef033

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95930 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100990 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46439 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97975 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15839 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23976 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73146 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30368 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98933 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11858 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86649 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53471 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11592 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4398 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45774 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81758 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103018 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22997 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16764 "Found 1 new test failure: imported/w3c/web-platform-tests/html/infrastructure/safe-passing-of-structured-data/shared-array-buffers/no-coop-coep.https.any.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82186 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23248 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82673 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81548 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20475 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26137 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3583 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16334 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22960 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28115 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22619 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26099 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24360 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1567,12 +1555,41 @@ void AXObjectCache::childrenChanged(Node* node, Element* changedChild)
 
+AccessibilityObject* AXObjectCache::getIncludingAncestors(RenderObject* renderer) const
+{
+    auto* current = renderer;
```
we could cut down lines of code with a for loop

twilco committed:
AX: Exiting early from AXObjectCache::childrenChanged(RenderObject*) when we can't `get` an object for anonymous renderers can result in content being missing from the accessibility tree
https://bugs.webkit.org/show_bug.cgi?id=290069
rdar://147283521

Reviewed by Chris Fleizach.

Exiting early from AXObjectCache::childrenChanged(RenderObject*) when we can't `AXObjectCache::get` an object for
anonymous renderers can result in content being missing from the accessibility tree. Sometimes the only children-changed
we get from the render tree for some subtrees is for an anonymous renderer, so if we just drop it, we can have a stale
accessibility tree. This problem is specific to anonymous renderers because we walk the DOM when building the accessibility
tree, so it's unlikely that we will have created an accessibility object for this renderer (since it only exists in the render tree).

This commit resolves the problem by changing AXObjectCache::childrenChanged(RenderObject*) to check whether a renderer
is anonymous, and if we can't AXObjectCache::get an accessibility object for it, ascends to the nearest thing we can get
an accessibility object for and submits the children-changed event for that instead.

This bug was found on a real tax preparation webpage, but unfortunately I couldn't figure out how to distill it into a layout test.
The key seems to be anonymous continuation renderers destroyed and recreated as part of a call to Node::insertBefore().
Newly added test dynamic-inline-continuation.html gets close to reproducing the issue, following many of the same
codepaths as the real bug, but unfortunately will still pass even without this commit.

For posterity, here's a stacktrace that gives us one of these children-changed events that used to be dropped but no
longer is:

2   AXObjectCache::childrenChanged(WebCore::RenderObject*, WebCore::RenderObject*)
3   RenderTreeBuilder::Block::detach(WebCore::RenderBlock&, WebCore::RenderObject&, WebCore::RenderTreeBuilder::WillBeDestroyed, WebCore::RenderTreeBuilder::CanCollapseAnonymousBlock)
4   RenderTreeBuilder::destroy(WebCore::RenderObject&, WebCore::RenderTreeBuilder::CanCollapseAnonymousBlock)
5   RenderTreeBuilder::destroyAndCleanUpAnonymousWrappers(WebCore::RenderObject&, WebCore::RenderElement const*)
6   RenderTreeUpdater::tearDownRenderers(WebCore::Element&, WebCore::RenderTreeUpdater::TeardownType, WebCore::RenderTreeBuilder&)
7   RenderTreeUpdater::tearDownRenderers(WebCore::Element&, WebCore::RenderTreeUpdater::TeardownType)
8   ContainerNode::removeBetween(WebCore::Node*, WebCore::Node*, WebCore::Node&)
9   ContainerNode::removeChild(WebCore::Node&)
10  ContainerNode::removeSelfOrChildNodesForInsertion(WebCore::Node&, WTF::Vector<WTF::Ref<WebCore::Node, WTF::RawPtrTraits<WebCore::Node>, WTF::DefaultRefDerefTraits<WebCore::Node>>, 11ul, WTF::CrashOnOverflow, 16ul, WTF::FastMalloc>&)
11  ContainerNode::appendChildWithoutPreInsertionValidityCheck(WebCore::Node&)
12  ContainerNode::insertBefore(WebCore::Node&, WTF::RefPtr<WebCore::Node, WTF::RawPtrTraits<WebCore::Node>, WTF::DefaultRefDerefTraits<WebCore::Node>>&&)

This commit also inlines AXObjectCache::get(RenderObject&) and AXObjectCache::(Widget&) for better performance.

* LayoutTests/accessibility/mac/dynamic-inline-continuation-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-inline-continuation.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::childrenChanged):
(WebCore::AXObjectCache::getIncludingAncestors const):
* Source/WebCore/accessibility/AXObjectCache.h:

Canonical link: https://commits.webkit.org/292433@main
